apps/prairielearn/src/ee/lib/validateHTML.ts

```diff
@@ -0,0 +1,616 @@
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertInt(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^\d+$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an integer, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is an float (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param errors The list of errors to add to.
+ */
+function assertFloat(tag: string, key: string, val: string, errors: string[]) {
+  if (!(/^(\d+)\.?(\d*)(e-\d+)?$/.test(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be an floating-point number, but value provided is "${val}"`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is in a list of possibilities (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertInChoices(
+  tag: string,
+  key: string,
+  val: string,
+  choices: string[],
+  errors: string[],
+) {
+  if (!(choices.includes(val) || /^\{\{.*\}\}$/.test(val))) {
+    errors.push(
+      `${tag}: value for attribute ${key} must be in ${choices}, but value provided is ${val}`,
+    );
+  }
+}
+
+/**
+ * Checks that the required property is a boolean (or property to be inserted) or adds an error to the provided list.
+ * @param tag The name of the tag being checked.
+ * @param key The property name.
+ * @param val The property value.
+ * @param choices The list of potential choices for the property.
+ * @param errors The list of errors to add to.
+ */
+function assertBool(tag: string, key: string, val: string, errors: string[]) {
+  assertInChoices(tag, key, val, ['true', 'false'], errors);
+}
+
+/**
+ * Checks that a tag has valid properties.
+ * @param ast The tree to consider, rooted at the tag.
+ * @returns The list of errors for the tag, if any.
+ */
+function checkTag(ast: any): string[] {
+  switch (ast.tagName) {
+    case 'pl-multiple-choice':
+      return checkMC(ast);
+    case 'pl-integer-input':
+      return checkInt(ast);
+    case 'pl-number-input':
+      return checkNum(ast);
+    case 'pl-answer':
+      return checkMCAnswer(ast);
+    case 'pl-overlay':
+      return checkOverlay(ast);
+    case 'pl-location':
+      return checkLocation(ast);
+    case 'pl-order-blocks':
+      return checkOrderBlocks(ast);
+    case 'pl-matrix-input':
+      return checkMatrix(ast);
+  }
+  return [];
+}
+
+/**
+ * Checks that a multiple choice question tag has valid properties.
```

Comment:
```suggestion
 * Checks that a `pl-multiple-choice` element has valid properties.
```

Reply:
Changed in 0e87e7e4333c2d4d8dfbc2116aad0c0acd81707b.